example/glue/EvacuatorDelegate.hpp

```diff
@@ -0,0 +1,245 @@
+	 * object that requires evacuation. If any of these objects are evacuated to nursery space, this method must
+	 * return true.
+	 */
+	bool scanIndirectObjects(omrobjectptr_t objectptr) { return false; }
+
+	/**
+	 * Iterate over root objects (eg, heap references held in thread slots when a gc cycle starts), calling
+	 * _evacuator->evacuateRootObject() for each root object. An optimzation (obviate a write barrier) may be
+	 * realized by calling _evacuator->evacuateThreadSlot() for references from the mutator stack, but this is
+	 * not necessary.
+	 */
+	void scanRoots();
+
+	/**
+	 * Reiterate over thread slots, calling _evacuator->rescanThreadSlot() for each thread slot. This is required
+	 * only for delegates that call _evacuator->evacuateThreadSlot() for stack-based heap references.
+	 */
+	void rescanThreadSlots() { }
+
+	/**
+	 * Some delegates may elect to hold back some root references until after the main root scanning and heap scan
+	 * (and thread rescanning, if used) are complete. This method is called after this point to determine whether
+	 * there are more roots to be cleared. If a value true is returned, scanClearable() will be called to allow the
+	 * next set of root objects to be cleared.
+	 */
+	bool hasClearable() { return !_isCleared; }
+
+	/**
+	 * If the return value from hasClearable() indicates that there are more roots to be cleared, this method will
+	 * be called. As for scanRoots(), _evacuator->evacuateRootObjerct() must be called each root object that is to
+	 * be cleared in the present clearing phase.
+	 *
+	 * Clearing may be partitioned into multiple phases. Each phase begins with a call to scanClearable() after
+	 * hasClearable() returns true. All objects reachable in the reference graph from the cleared objects are then
+	 * evacuated as required. This repeats until hasClearable() returns false.
+	 *
+	 */
+	void scanClearable();
+
+	/**
+	 * This is called before a gc thread enters a wait state (wits for work), to give the delgate an opportunity to
+	 * flush any material that it might be holding that may be relevant to other threads. Implementation is entirely
+	 * language-specific and may not be required.
+	 */
+	void flushForWaitState() { }
+
+	/**
+	 * This is called after a gc thread completes the last clearing phase, to give the delgate an opportunity to
+	 * flush any material that it might be holding that may be relevant to other threads. Implementation is entirely
+	 * language-specific and may not be required.
+	 */
+	void flushForEndCycle() { }
+
+	MM_EvacuatorDelegate()
+		: _env(NULL)
+		, _controller(NULL)
+		, _evacuator(NULL)
+		, _isCleared(false)
+	{ }
+
+#if defined(EVACUATOR_DEBUG_DELEGATE)
+	void debugValidateObject(omrobjectptr_t objectptr) { }
+	void debugValidateObject(MM_ForwardedHeader *forwardedHeader) { }
+	const char *
+	debugGetClassname(omrobjectptr_t objectptr, char *buffer, uintptr_t bufferLength)
```

Comment:
What is the debugGetClassName used for?

Reply:
This used used with tracing (EVACUATOR_DEBUG) object copying. Classname and size are printed. This is never enabled in submitted evacuator PRs.